sdk/nodejs/runtime/closure/package.ts

```diff
@@ -0,0 +1,267 @@
+                last = curPath;
+            } catch (e) {
+                // current path is not a module
+            }
+            directories.pop();
+        }
+        if (last === undefined || lastFullPath === undefined) {
+            throw new Error(`no package.json found for ${path}`);
+        }
+        const packageDefinitionAbsPath = lastFullPath.slice(0, lastFullPath.indexOf(last)) + last + "/package.json";
+        return require(packageDefinitionAbsPath);
+    } catch (err) {
+        return undefined;
+    }
+}
+
+// a module's implementations are leaves of the document tree.
+function getAllLeafStrings(objectOrPath: SubExports, opts?: RequireOpts): string[] {
+    if (objectOrPath === null) {
+        // module blacklisted return no implementations
+        return [];
+    }
+    if (typeof objectOrPath === "string") {
+        return [objectOrPath];
+    }
+    const strings: string[] = [];
+    for (const [key, value] of Object.entries(objectOrPath)) {
+        if (opts && !opts.isRequire && key === "require") {
+            continue;
+        }
+        if (opts && !opts.isImport && key === "import") {
+            continue;
+        }
+        const leaves = getAllLeafStrings(value);
+        if (leaves === []) {
+            // if there's an environment where this export does not work,
+            // don't suggest requires from this match as a more preferable path may
+            // match this file.
+            return [];
+        }
+        strings.push(...leaves);
+    }
+    return strings;
+}
+
+// from https://github.com/nodejs/node/blob/b191e66ddf/lib/internal/modules/esm/resolve.js#L686
+function patternKeyCompare(a: string, b: string) {
+    const aPatternIndex = a.indexOf("*");
+    const bPatternIndex = b.indexOf("*");
+    const baseLenA = aPatternIndex === -1 ? a.length : aPatternIndex + 1;
+    const baseLenB = bPatternIndex === -1 ? b.length : bPatternIndex + 1;
+    if (baseLenA > baseLenB) {
+        return -1;
+    }
+    if (baseLenB > baseLenA) {
+        return 1;
+    }
+    if (aPatternIndex === -1) {
+        return 1;
+    }
+    if (bPatternIndex === -1) {
+        return -1;
+    }
+    if (a.length > b.length) {
+        return -1;
+    }
+    if (b.length > a.length) {
+        return 1;
+    }
+    return 0;
+}
+
+type SrcPrefix = string;
+type Rule = [SrcPrefix, {
+    modPrefix: string;
+    modSuffix: string;
+    srcSuffix: string;
+}];
+
+function makeRule(srcPattern: string, modPattern: string): Rule {
+    const srcSplit = srcPattern.split("*"); // NodeJS doesn't error out when provided multiple '*'.
+    const modSplit = modPattern.split("*");
+    if (srcSplit.length > 2 || modSplit.length > 2) {
+        // there is undefined behavior on more than 1 "*"
+        // see https://github.com/nodejs/node/blob/b191e66ddf/lib/internal/modules/esm/resolve.js#L664
+        throw new Error("multiple wildcards in single export target specification");
```

Comment:
Do these diagnostic messages go to the end user somehow, if so is there diagnostics about the particular import - I'm guessing that's either srcPattern or modPattern - so the user knows which module misbehaved?

Reply:
I'm not sure if something catches these errors. I could extract the module name by splitting the as an argument to throw a better exception internally, but I've hit my time allowance.

the exports name will refuse to wildcard match if the original string has more than 1 asterisk.
it's a string replacement in the algorithm so if the target has multiple asterisks, each asterisk would be replaced with the match.

I think it's generally a never condition given how impractical structuring your files would be in a repository like this.